projects/igniteui-angular/src/lib/grids/tree-grid/tree-grid.component.ts

```diff
@@ -720,8 +720,23 @@ export class IgxTreeGridComponent extends IgxGridBaseDirective implements GridTy
         //  if this is flat self-referencing data, and CascadeOnDelete is set to true
         //  and if we have transactions we should start pending transaction. This allows
         //  us in case of delete action to delete all child rows as single undo action
-        return this.gridAPI.deleteRowById(rowId);
+        const args = {
+            rowID: rowId,
+            cancel: false,
+            rowData: this.getRowData(rowId),
+            oldValue: null
+        };
+        this.rowDelete.emit(args);
+        if (args.cancel) {
+            return;
+        }
 
+        const record = this.gridAPI.deleteRowById(rowId);
+        if (record !== null && record !== undefined) {
+            //  TODO: should we emit this when cascadeOnDelete is true for each row?!?!
```

Comment:
We shouldn't merge changes with TODOs

I'd say we should probably fire the event for all rows when cascading deletes them. @damyanpetev , what do you think?

Reply:
I think it'll be a bit excessive firing a potentially pretty large amount of events for a single user interaction. For the most part I'm assuming the cascade delete reflects the default behavior of the data source (backend/DB) and the root record deletion is enough information considering the user is the one that configured the cascade so it shouldn't be a surprise.
What could be useful actually, since `rowData` may not be hierarchical is to also provide the public row interface in the arguments - in this case the `IgxTreeGridRow` which will also give access to the `children` should the data not contain those. So even if the cascade is only in the client grid, it shouldn't be much of a hassle to grab all the children

Reply:
So after a quick discussion, we've settled on this: The delete event is fired once delete on flat data using primary and foreign keys - that keeps it consistent with the equivalent interaction on actual hierarchical data and also with cascade on (default) the presumption is the client grid is mimicking the behavior of the backend/DB. Should there be need to access the child structure under a deleted row - my initial suggestion to add the row interface to all events was a bit off, as after a delete the grid APIs no longer make sense and the event interfaces won't accommodate that easily ATM which is something we might tackle on down the road.
The children should still be available in the row data (if hierarchical) or should be accessible based on the row key from the grid API (flat data). The latter applies similarly to access an added row and its parent structure.